routing/src/vrf.rs

```diff
@@ -0,0 +1,447 @@
+    pub shim: VrfId,
+    pub rc: Rc<Nhop>,
+}
+
+#[allow(unused)]
+pub struct Vrf {
+    #[allow(unused)]
+    pub name: String,
+    pub vrfid: VrfId,
+    pub(crate) routesv4: RTrieMap<Ipv4Prefix, Route>,
+    pub(crate) routesv6: RTrieMap<Ipv6Prefix, Route>,
+    pub(crate) nhstore: NhopStore,
+    pub(crate) vni: Option<Vni>,
+}
+
+#[allow(dead_code)]
+impl Vrf {
+    /////////////////////////////////////////////////////////////////////////
+    /// Create a new VRF with the given name and vrfId
+    /// Initial capacities are 0. We may want to have some default sizes.
+    /// Else, we can always call Self::with_capacities().
+    /////////////////////////////////////////////////////////////////////////
+    pub fn new(name: &str, vrfid: VrfId) -> Self {
+        Self::with_capacities(name, vrfid, 0, 0)
+    }
+
+    /////////////////////////////////////////////////////////////////////////
+    /// Create VRF with some initial capacities
+    /////////////////////////////////////////////////////////////////////////
+    pub fn with_capacities(name: &str, vrfid: VrfId, capa_v4: usize, capa_v6: usize) -> Self {
+        let mut vrf = Self {
+            name: name.to_owned(),
+            vrfid,
+            routesv4: RTrieMap::with_capacity(capa_v4),
+            routesv6: RTrieMap::with_capacity(capa_v6),
+            nhstore: NhopStore::new(),
+            vni: None,
+        };
+        /* add default routes with default next-hop with action DROP */
+        vrf.add_route(
+            &Prefix::root_v4(),
+            Route::default(),
+            &[RouteNhop::default()],
+        );
+        vrf.add_route(
+            &Prefix::root_v6(),
+            Route::default(),
+            &[RouteNhop::default()],
+        );
+        vrf
+    }
+
+    /////////////////////////////////////////////////////////////////////////
+    /// Set the VNI for a Vrf
+    /////////////////////////////////////////////////////////////////////////
+    pub fn set_vni(&mut self, vni: Vni) {
+        self.vni = Some(vni);
+    }
+
+    #[inline(always)]
+    #[must_use]
+    fn register_shared_nhop(&mut self, nhop: &RouteNhop) -> Rc<Nhop> {
+        self.nhstore.add_nhop(&nhop.key)
+    }
+
+    /////////////////////////////////////////////////////////////////////////
+    /// Register a shared next-hop for the route if not there
+    /////////////////////////////////////////////////////////////////////////
+    fn register_shared_nhops(&mut self, route: &mut Route, nhops: &[RouteNhop]) {
+        for nhop in nhops {
+            // shim next-hop created here
+            let shared = self.register_shared_nhop(nhop);
+            let shim = ShimNhop {
+                shim: nhop.vrfid,
+                rc: shared,
+            };
+            route.s_nhops.push(shim);
+        }
+    }
+
+    #[inline(always)]
+    fn deregister_shared_nhop(&mut self, shim: ShimNhop) {
+        let key = shim.rc.key;
+        drop(shim);
+        self.nhstore.del_nhop(&key);
```

Comment:
why the manual drop?  It will drop when it goes out of scope at the end of the function

Reply:
Because we want to decrement the refcount _before_ we call self.nhstore.del_nhop() in order to keep stale next-hops around. If we would let it drop by itself, that would happen after the call, right?

Reply:
ah, that makes sense.